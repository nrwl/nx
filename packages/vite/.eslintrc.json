{
  "extends": ["../../.eslintrc.json"],
  "ignorePatterns": ["!**/*"],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "rules": {}
    },
    {
      "files": ["*.ts", "*.tsx"],
      "rules": {}
    },
    {
      "files": ["*.js", "*.jsx"],
      "rules": {}
    },
    {
      "files": ["./package.json", "./generators.json", "./executors.json"],
      "parser": "jsonc-eslint-parser",
      "rules": {
        "@nx/nx-plugin-checks": "error"
      }
    },
    {
      "files": ["./package.json"],
      "parser": "jsonc-eslint-parser",
      "rules": {
        "@nx/dependency-checks": [
          "error",
          {
            "buildTargets": ["build-base"],
            "ignoredDependencies": [
              "nx",
              "typescript",
              "vite",
              // we only check if the package is installed
              "eslint",
              // we ensure it is installed and only use it when eslint is installed
              "@nx/eslint"
            ]
          }
        ]
      }
    }
  ]
}
