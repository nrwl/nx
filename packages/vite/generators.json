{
  "name": "Nx Vite",
  "version": "0.1",
  "generators": {
    "init": {
      "factory": "./src/generators/init/init#initGeneratorInternal",
      "schema": "./src/generators/init/schema.json",
      "description": "Initialize Vite in the workspace.",
      "aliases": ["ng-add"],
      "hidden": true
    },
    "configuration": {
      "factory": "./src/generators/configuration/configuration#viteConfigurationGeneratorInternal",
      "schema": "./src/generators/configuration/schema.json",
      "description": "Add Vite configuration to an application.",
      "aliases": ["config"],
      "hidden": false
    },
    "setup-paths-plugin": {
      "factory": "./src/generators/setup-paths-plugin/setup-paths-plugin",
      "schema": "./src/generators/setup-paths-plugin/schema.json",
      "description": "Sets up the nxViteTsPaths plugin to enable support for workspace libraries."
    },
    "convert-to-inferred": {
      "factory": "./src/generators/convert-to-inferred/convert-to-inferred",
      "schema": "./src/generators/convert-to-inferred/schema.json",
      "description": "Convert existing Vite project(s) using `@nx/vite:*` executors to use `@nx/vite/plugin`. Defaults to migrating all projects. Pass '--project' to migrate only one target."
    },
    "vitest": {
      "factory": "./src/generators/vitest/vitest-generator",
      "schema": "./src/generators/vitest/schema.json",
      "description": "Generate a vitest configuration."
    }
  }
}
