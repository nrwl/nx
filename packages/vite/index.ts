export * from './src/utils/versions';
export * from './src/utils/generator-utils';
export * from './src/utils/e2e-web-server-info-utils';
export { type ViteConfigurationGeneratorSchema } from './src/generators/configuration/schema';
export { viteConfigurationGenerator } from './src/generators/configuration/configuration';
export { type VitestGeneratorSchema } from './src/generators/vitest/schema';
export { vitestGenerator } from './src/generators/vitest/vitest-generator';
export { type InitGeneratorSchema } from './src/generators/init/schema';
export { initGenerator } from './src/generators/init/init';
