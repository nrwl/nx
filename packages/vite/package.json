{
  "name": "@nx/vite",
  "version": "0.0.1",
  "private": false,
  "description": "The Nx Plugin for building and testing applications using Vite",
  "repository": {
    "type": "git",
    "url": "https://github.com/nrwl/nx.git",
    "directory": "packages/vite"
  },
  "keywords": [
    "Monorepo",
    "Vite",
    "Web",
    "CLI",
    "Front-end"
  ],
  "main": "./index",
  "typings": "./index.d.ts",
  "author": "Victor Savkin",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/nrwl/nx/issues"
  },
  "homepage": "https://nx.dev",
  "generators": "./generators.json",
  "executors": "./executors.json",
  "ng-update": {
    "requirements": {},
    "migrations": "./migrations.json"
  },
  "dependencies": {
    "@nx/devkit": "file:../devkit",
    "@phenomnomnominal/tsquery": "~5.0.1",
    "@swc/helpers": "~0.5.0",
    "enquirer": "~2.3.6",
    "@nx/js": "file:../js",
    "tsconfig-paths": "^4.1.2",
    "minimatch": "9.0.3",
    "semver": "^7.6.3"
  },
  "peerDependencies": {
    "vite": "^5.0.0 || ^6.0.0",
    "vitest": "^1.3.1 || ^2.0.0 || ^3.0.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "exports": {
    ".": "./index.js",
    "./plugin": "./plugin.js",
    "./package.json": "./package.json",
    "./migrations.json": "./migrations.json",
    "./generators.json": "./generators.json",
    "./executors.json": "./executors.json",
    "./executors": "./executors.js",
    "./src/executors/*/schema.json": "./src/executors/*/schema.json",
    "./src/executors/*.impl": "./src/executors/*.impl.js",
    "./src/executors/*/compat": "./src/executors/*/compat.js",
    "./plugins/nx-copy-assets.plugin": "./plugins/nx-copy-assets.plugin.js",
    "./plugins/nx-tsconfig-paths.plugin": "./plugins/nx-tsconfig-paths.plugin.js",
    "./plugins/rollup-replace-files.plugin": "./plugins/rollup-replace-files.plugin.js"
  }
}
