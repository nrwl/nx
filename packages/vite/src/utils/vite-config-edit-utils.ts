import { applyChangesToString, ChangeType, Tree } from '@nx/devkit';
import { findNodes } from '@nx/js';
import { TargetFlags } from './generator-utils';
import type {
  ArrayLiteralExpression,
  Node,
  PropertyAssignment,
  ReturnStatement,
} from 'typescript';

export function ensureViteConfigIsCorrect(
  tree: Tree,
  path: string,
  buildConfigString: string,
  buildConfigObject: {},
  imports: string[],
  plugins: string[],
  testConfigString: string,
  testConfigObject: {},
  cacheDir: string,
  projectAlreadyHasViteTargets?: TargetFlags
): boolean {
  const fileContent = tree.read(path, 'utf-8');

  let updatedContent = undefined;

  if (!projectAlreadyHasViteTargets?.test && testConfigString?.length) {
    updatedContent = handleBuildOrTestNode(
      fileContent,
      testConfigString,
      testConfigObject,
      'test'
    );
  }

  if (!projectAlreadyHasViteTargets?.build && buildConfigString?.length) {
    updatedContent = handleBuildOrTestNode(
      updatedContent ?? fileContent,
      buildConfigString,
      buildConfigObject,
      'build'
    );
  }

  updatedContent =
    handlePluginNode(updatedContent ?? fileContent, imports, plugins) ??
    updatedContent;

  if (cacheDir?.length) {
    updatedContent = handleCacheDirNode(
      updatedContent ?? fileContent,
      cacheDir
    );
  }

  if (updatedContent) {
    tree.write(path, updatedContent);
    return true;
  } else {
    return false;
  }
}

function handleBuildOrTestNode(
  updatedFileContent: string,
  configContentString: string,
  configContentObject: {},
  name: 'build' | 'test'
): string | undefined {
  const { tsquery } = require('@phenomnomnominal/tsquery');
  const buildOrTestNode = tsquery.query(
    updatedFileContent,
    `PropertyAssignment:has(Identifier[name="${name}"])`
  );

  if (buildOrTestNode.length) {
    return tsquery.replace(
      updatedFileContent,
      `PropertyAssignment:has(Identifier[name="${name}"])`,
      (node: PropertyAssignment) => {
        const existingProperties = tsquery.query(
          node.initializer,
          'PropertyAssignment'
        ) as PropertyAssignment[];
        let updatedPropsString = '';
        for (const prop of existingProperties) {
          const propName = prop.name.getText();
          if (
            !configContentObject[propName] &&
            propName !== 'dir' &&
            propName !== 'reportsDirectory' &&
            propName !== 'provider'
          ) {
            // NOTE: Watch for formatting.
            updatedPropsString += `    '${propName}': ${prop.initializer.getText()},\n`;
          }
        }
        for (const [propName, propValue] of Object.entries(
          configContentObject
        )) {
          // NOTE: Watch for formatting.
          if (propName === 'coverage') {
            let propString = `    '${propName}': {\n`;
            for (const [pName, pValue] of Object.entries(propValue)) {
              if (pName === 'provider') {
                propString += `    '${pName}': ${pValue} as const,\n`;
              } else {
                propString += `    '${pName}': '${pValue}',\n`;
              }
            }
            propString += `}`;
            updatedPropsString += `${propString}\n`;
          } else {
            updatedPropsString += `    '${propName}': ${JSON.stringify(
              propValue
            )},\n`;
          }
        }
        return `${name}: {
${updatedPropsString}  }`;
      }
    );
  } else {
    const foundDefineConfig = tsquery.query(
      updatedFileContent,
      'CallExpression:has(Identifier[name="defineConfig"])'
    );

    if (foundDefineConfig.length) {
      const conditionalConfig = tsquery.query(
        foundDefineConfig[0],
        'ArrowFunction'
      );

      if (conditionalConfig.length) {
        if (name === 'build') {
          return transformConditionalConfig(
            conditionalConfig,
            updatedFileContent,
            configContentString
          );
        } else {
          // no test config in conditional config
          return updatedFileContent;
        }
      } else {
        const propertyAssignments = tsquery.query(
          foundDefineConfig[0],
          'PropertyAssignment'
        );

        if (propertyAssignments.length) {
          return applyChangesToString(updatedFileContent, [
            {
              type: ChangeType.Insert,
              index: propertyAssignments[0].getStart(),
              text: configContentString,
            },
          ]);
        } else {
          return applyChangesToString(updatedFileContent, [
            {
              type: ChangeType.Insert,
              index: foundDefineConfig[0].getStart() + 14,
              text: configContentString,
            },
          ]);
        }
      }
    } else {
      // build config does not exist and defineConfig is not used
      // could also potentially be invalid syntax, so try-catch
      try {
        const defaultExport = tsquery.query(
          updatedFileContent,
          'ExportAssignment'
        );
        const found = tsquery.query(
          defaultExport?.[0],
          'ObjectLiteralExpression'
        );
        const startOfObject = found?.[0].getStart();
        return applyChangesToString(updatedFileContent, [
          {
            type: ChangeType.Insert,
            index: startOfObject + 1,
            text: configContentString,
          },
        ]);
      } catch {
        return updatedFileContent;
      }
    }
  }
}

function transformCurrentBuildObject(
  index: number,
  returnStatements: ReturnStatement[],
  appFileContent: string,
  buildConfigObject: {}
): string | undefined {
  if (!returnStatements?.[index]) {
    return undefined;
  }
  const { tsquery } = require('@phenomnomnominal/tsquery');
  const currentBuildObject = tsquery
    .query(returnStatements[index], 'ObjectLiteralExpression')?.[0]
    .getText();

  const currentBuildObjectStart = returnStatements[index].getStart();
  const currentBuildObjectEnd = returnStatements[index].getEnd();
  const newReturnObject = tsquery.replace(
    returnStatements[index].getText(),
    'ObjectLiteralExpression',
    (_node: Node) => {
      return `{
        ...${currentBuildObject},
        ...${JSON.stringify(buildConfigObject)}
      }`;
    }
  );

  const newContents = applyChangesToString(appFileContent, [
    {
      type: ChangeType.Delete,
      start: currentBuildObjectStart,
      length: currentBuildObjectEnd - currentBuildObjectStart,
    },
    {
      type: ChangeType.Insert,
      index: currentBuildObjectStart,
      text: newReturnObject,
    },
  ]);

  return newContents;
}

function transformConditionalConfig(
  conditionalConfig: Node[],
  appFileContent: string,
  buildConfigObject: {}
): string | undefined {
  const { tsquery } = require('@phenomnomnominal/tsquery');
  const { SyntaxKind } = require('typescript');
  const functionBlock = tsquery.query(conditionalConfig[0], 'Block');
  const ifStatement = tsquery.query(functionBlock?.[0], 'IfStatement');

  const binaryExpressions = tsquery.query(ifStatement?.[0], 'BinaryExpression');

  const buildExists = binaryExpressions?.find(
    (binaryExpression) => binaryExpression.getText() === `command === 'build'`
  );

  const buildExistsExpressionIndex = binaryExpressions?.findIndex(
    (binaryExpression) => binaryExpression.getText() === `command === 'build'`
  );

  const serveExists = binaryExpressions?.find(
    (binaryExpression) => binaryExpression.getText() === `command === 'serve'`
  );

  const elseKeywordExists = findNodes(ifStatement?.[0], SyntaxKind.ElseKeyword);
  const returnStatements: ReturnStatement[] = tsquery.query(
    ifStatement[0],
    'ReturnStatement'
  );

  if (!buildExists) {
    if (serveExists && elseKeywordExists) {
      // build options live inside the else block
      return (
        transformCurrentBuildObject(
          returnStatements?.length - 1,
          returnStatements,
          appFileContent,
          buildConfigObject
        ) ?? appFileContent
      );
    } else {
      // no build options exist yet
      const functionBlockStart = functionBlock?.[0].getStart();
      const newContents = applyChangesToString(appFileContent, [
        {
          type: ChangeType.Insert,
          index: functionBlockStart + 1,
          text: `
            if (command === 'build') {
              return ${JSON.stringify(buildConfigObject)}
            }
            `,
        },
      ]);
      return newContents;
    }
  } else {
    // build already exists
    // it will be the return statement which lives
    // at the buildExistsExpressionIndex

    return (
      transformCurrentBuildObject(
        buildExistsExpressionIndex,
        returnStatements,
        appFileContent,
        buildConfigObject
      ) ?? appFileContent
    );
  }
}

function handlePluginNode(
  appFileContent: string,
  imports: string[],
  plugins: string[]
): string | undefined {
  const { tsquery } = require('@phenomnomnominal/tsquery');
  const file = tsquery.ast(appFileContent);
  const pluginsNode = tsquery.query(
    file,
    'PropertyAssignment:has(Identifier[name="plugins"])'
  );

  let writeFile = false;

  if (pluginsNode.length) {
    appFileContent = tsquery.replace(
      file.getText(),
      'PropertyAssignment:has(Identifier[name="plugins"])',
      (node: Node) => {
        const found = tsquery.query(
          node,
          'ArrayLiteralExpression'
        ) as ArrayLiteralExpression[];
        let updatedPluginsString = '';

        const existingPluginNodes = found?.[0].elements ?? [];

        for (const plugin of existingPluginNodes) {
          updatedPluginsString += `${plugin.getText()}, `;
        }

        for (const plugin of plugins) {
          if (
            !existingPluginNodes?.some((node) =>
              node.getText().includes(plugin)
            )
          ) {
            updatedPluginsString += `${plugin}, `;
          }
        }

        return `plugins: [${updatedPluginsString}]`;
      }
    );
    writeFile = true;
  } else {
    // Plugins node does not exist yet
    // So make one from scratch

    const foundDefineConfig = tsquery.query(
      file,
      'CallExpression:has(Identifier[name="defineConfig"])'
    );

    if (foundDefineConfig.length) {
      const conditionalConfig = tsquery.query(
        foundDefineConfig[0],
        'ArrowFunction'
      );

      if (conditionalConfig.length) {
        // We are NOT transforming the conditional config
        // with plugins
        writeFile = false;
      } else {
        const propertyAssignments = tsquery.query(
          foundDefineConfig[0],
          'PropertyAssignment'
        );

        if (propertyAssignments.length) {
          appFileContent = applyChangesToString(appFileContent, [
            {
              type: ChangeType.Insert,
              index: propertyAssignments[0].getStart(),
              text: `plugins: [${plugins.join(', ')}],`,
            },
          ]);
          writeFile = true;
        } else {
          appFileContent = applyChangesToString(appFileContent, [
            {
              type: ChangeType.Insert,
              index: foundDefineConfig[0].getStart() + 14,
              text: `plugins: [${plugins.join(', ')}],`,
            },
          ]);
          writeFile = true;
        }
      }
    } else {
      // Plugins option does not exist and defineConfig is not used
      // could also potentially be invalid syntax, so try-catch
      try {
        const defaultExport = tsquery.query(file, 'ExportAssignment');
        const found = tsquery?.query(
          defaultExport?.[0],
          'ObjectLiteralExpression'
        );
        const startOfObject = found?.[0].getStart();
        appFileContent = applyChangesToString(appFileContent, [
          {
            type: ChangeType.Insert,
            index: startOfObject + 1,
            text: `plugins: [${plugins.join(', ')}],`,
          },
        ]);
        writeFile = true;
      } catch {
        writeFile = false;
      }
    }
  }
  if (writeFile) {
    const filteredImports = filterImport(appFileContent, imports);
    return filteredImports.join(';\n') + '\n' + appFileContent;
  }
}

function filterImport(appFileContent: string, imports: string[]): string[] {
  const { tsquery } = require('@phenomnomnominal/tsquery');
  const file = tsquery.ast(appFileContent);
  const importNodes = tsquery.query(
    file,
    ':matches(ImportDeclaration, VariableStatement)'
  );

  const importsArrayExisting = importNodes?.map((node) => {
    return node.getText().slice(0, -1);
  });

  return imports.filter((importString) => {
    return !importsArrayExisting?.includes(importString);
  });
}

function handleCacheDirNode(appFileContent: string, cacheDir: string): string {
  const { tsquery } = require('@phenomnomnominal/tsquery');

  const file = tsquery.ast(appFileContent);
  const cacheDirNode = tsquery.query(
    file,
    'PropertyAssignment:has(Identifier[name="cacheDir"])'
  );

  if (!cacheDirNode?.length || cacheDirNode?.length === 0) {
    // cacheDir node does not exist yet
    // So make one from scratch

    const foundDefineConfig = tsquery.query(
      file,
      'CallExpression:has(Identifier[name="defineConfig"])'
    );

    if (foundDefineConfig.length) {
      const conditionalConfig = tsquery.query(
        foundDefineConfig[0],
        'ArrowFunction'
      );

      if (conditionalConfig.length) {
        // We are NOT transforming the conditional config
        // with cacheDir
      } else {
        const propertyAssignments = tsquery.query(
          foundDefineConfig[0],
          'PropertyAssignment'
        );

        if (propertyAssignments.length) {
          appFileContent = applyChangesToString(appFileContent, [
            {
              type: ChangeType.Insert,
              index: propertyAssignments[0].getStart(),
              text: cacheDir,
            },
          ]);
        } else {
          appFileContent = applyChangesToString(appFileContent, [
            {
              type: ChangeType.Insert,
              index: foundDefineConfig[0].getStart() + 14,
              text: cacheDir,
            },
          ]);
        }
      }
    } else {
      // cacheDir option does not exist and defineConfig is not used
      // could also potentially be invalid syntax, so try-catch
      try {
        const defaultExport = tsquery.query(file, 'ExportAssignment');
        const found = tsquery?.query(
          defaultExport?.[0],
          'ObjectLiteralExpression'
        );
        const startOfObject = found?.[0].getStart();
        appFileContent = applyChangesToString(appFileContent, [
          {
            type: ChangeType.Insert,
            index: startOfObject + 1,
            text: cacheDir,
          },
        ]);
      } catch {}
    }
  }

  return appFileContent;
}
