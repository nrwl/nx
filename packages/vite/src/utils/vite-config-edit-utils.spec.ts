import { Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { tsquery } from '@phenomnomnominal/tsquery';
import {
  buildOption,
  buildOptionObject,
  conditionalConfig,
  configNoDefineConfig,
  imports,
  hasEverything,
  noBuildOptions,
  noBuildOptionsHasTestOption,
  noContentDefineConfig,
  plugins,
  someBuildOptions,
  someBuildOptionsSomeTestOption,
  testOption,
  testOptionObject,
} from './test-files/test-vite-configs';
import { ensureViteConfigIsCorrect } from './vite-config-edit-utils';

describe('ensureViteConfigIsCorrect', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  });

  it("should add build options if build options don't exist", () => {
    tree.write('apps/my-app/vite.config.ts', noBuildOptions);
    ensureViteConfigIsCorrect(
      tree,
      'apps/my-app/vite.config.ts',
      buildOption,
      buildOptionObject,
      imports,
      plugins,
      testOption,
      testOptionObject,
      '',
      { build: false, test: true, serve: false }
    );
    const appFileContent = tree.read('apps/my-app/vite.config.ts', 'utf-8');
    const file = tsquery.ast(appFileContent);
    const buildNode = tsquery.query(
      file,
      'PropertyAssignment:has(Identifier[name="build"])'
    );
    expect(buildNode).toBeDefined();
    expect(appFileContent).toMatchInlineSnapshot(`
      "import dts from 'vite-plugin-dts';
      import { joinPathFragments } from '@nx/devkit'
      import { defineConfig } from 'vite';
          import react from '@vitejs/plugin-react';
          import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';

          export default defineConfig({
            
          // Configuration for building your library.
          // See: https://vitejs.dev/guide/build.html#library-mode
          build: {
            lib: {
              // Could also be a dictionary or array of multiple entry points.
              entry: 'src/index.ts',
              name: 'my-app',
              fileName: 'index',
              // Change this to the formats you want to support.
              // Don't forget to update your package.json as well.
              formats: ['es']
            },
            rollupOptions: {
              // External packages that should not be bundled into your library.
              external: ['react', 'react-dom', 'react/jsx-runtime']
            }
          },cacheDir: '../../node_modules/.vitest',
            plugins: [react(), nxViteTsPaths(), ],

            test: {
              globals: true,
              environment: 'jsdom',
              include: ['src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],
            },

          });
          "
    `);
  });

  it('should add new build options if some build options already exist', () => {
    tree.write('apps/my-app/vite.config.ts', someBuildOptions);
    ensureViteConfigIsCorrect(
      tree,
      'apps/my-app/vite.config.ts',
      buildOption,
      buildOptionObject,
      imports,
      plugins,
      testOption,
      testOptionObject,
      '',
      { build: false, test: true, serve: false }
    );
    const appFileContent = tree.read('apps/my-app/vite.config.ts', 'utf-8');
    const file = tsquery.ast(appFileContent);
    const buildNode = tsquery.query(
      file,
      'PropertyAssignment:has(Identifier[name="build"])'
    );
    expect(buildNode).toBeDefined();
    expect(appFileContent).toMatchInlineSnapshot(`
      "import dts from 'vite-plugin-dts';
      import { joinPathFragments } from '@nx/devkit'
      import { defineConfig } from 'vite';
          import react from '@vitejs/plugin-react';
          import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';

          export default defineConfig({
            cacheDir: '../../node_modules/.vitest',
            plugins: [react(), nxViteTsPaths(), ],

            test: {
              globals: true,
              environment: 'jsdom',
              include: ['src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],
            },

            build: {
          'my': 'option',
          'lib': {"entry":"src/index.ts","name":"my-app","fileName":"index","formats":["es"]},
          'rollupOptions': {"external":["react","react-dom","react/jsx-runtime"]},
        }

          });
          "
    `);
  });

  it('should add build and test options if defineConfig is empty', () => {
    tree.write('apps/my-app/vite.config.ts', noContentDefineConfig);
    ensureViteConfigIsCorrect(
      tree,
      'apps/my-app/vite.config.ts',
      buildOption,
      buildOptionObject,
      imports,
      plugins,
      testOption,
      testOptionObject,
      '',
      { build: false, test: false, serve: false }
    );
    const appFileContent = tree.read('apps/my-app/vite.config.ts', 'utf-8');
    const file = tsquery.ast(appFileContent);
    const buildNode = tsquery.query(
      file,
      'PropertyAssignment:has(Identifier[name="build"])'
    );
    expect(buildNode).toBeDefined();
    expect(appFileContent).toMatchInlineSnapshot(`
      "import dts from 'vite-plugin-dts';
      import { joinPathFragments } from '@nx/devkit'

          /// <reference types="vitest" />
          import { defineConfig } from 'vite';
          import react from '@vitejs/plugin-react';
          import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';

          export default defineConfig({
          // Configuration for building your library.
          // See: https://vitejs.dev/guide/build.html#library-mode
          plugins: [react(), nxViteTsPaths()],build: {
            lib: {
              // Could also be a dictionary or array of multiple entry points.
              entry: 'src/index.ts',
              name: 'my-app',
              fileName: 'index',
              // Change this to the formats you want to support.
              // Don't forget to update your package.json as well.
              formats: ['es']
            },
            rollupOptions: {
              // External packages that should not be bundled into your library.
              external: ['react', 'react-dom', 'react/jsx-runtime']
            }
          },test: {
              globals: true,
              environment: 'jsdom',
              include: ['src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],
          },});
          "
    `);
  });

  it('should add build options if it is using conditional config - do nothing for test', () => {
    tree.write('apps/my-app/vite.config.ts', conditionalConfig);
    ensureViteConfigIsCorrect(
      tree,
      'apps/my-app/vite.config.ts',
      buildOption,
      buildOptionObject,
      imports,
      plugins,
      testOption,
      testOptionObject,
      '',
      { build: false, test: false, serve: false }
    );
    const appFileContent = tree.read('apps/my-app/vite.config.ts', 'utf-8');
    const file = tsquery.ast(appFileContent);
    const buildNode = tsquery.query(
      file,
      'PropertyAssignment:has(Identifier[name="build"])'
    );
    expect(buildNode).toBeDefined();
    expect(appFileContent).toMatchSnapshot();
  });

  it('should add build options if defineConfig is not used', () => {
    tree.write('apps/my-app/vite.config.ts', configNoDefineConfig);
    ensureViteConfigIsCorrect(
      tree,
      'apps/my-app/vite.config.ts',
      buildOption,
      buildOptionObject,
      imports,
      plugins,
      testOption,
      testOptionObject,
      '',
      { build: false, test: false, serve: false }
    );
    const appFileContent = tree.read('apps/my-app/vite.config.ts', 'utf-8');
    const file = tsquery.ast(appFileContent);
    const buildNode = tsquery.query(
      file,
      'PropertyAssignment:has(Identifier[name="build"])'
    );
    expect(buildNode).toBeDefined();
    expect(appFileContent).toMatchInlineSnapshot(`
      "import dts from 'vite-plugin-dts';
      import { joinPathFragments } from '@nx/devkit'
      import { defineConfig } from 'vite';
          import react from '@vitejs/plugin-react';
          import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';

          export default {
          // Configuration for building your library.
          // See: https://vitejs.dev/guide/build.html#library-mode
          build: {
            lib: {
              // Could also be a dictionary or array of multiple entry points.
              entry: 'src/index.ts',
              name: 'my-app',
              fileName: 'index',
              // Change this to the formats you want to support.
              // Don't forget to update your package.json as well.
              formats: ['es']
            },
            rollupOptions: {
              // External packages that should not be bundled into your library.
              external: ['react', 'react-dom', 'react/jsx-runtime']
            }
          },test: {
              globals: true,
              environment: 'jsdom',
              include: ['src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],
          },
            plugins: [react(), nxViteTsPaths(), ],
          };
          "
    `);
  });

  it('should not do anything if cannot understand syntax of vite config', () => {
    tree.write('apps/my-app/vite.config.ts', `console.log('Unknown syntax')`);
    ensureViteConfigIsCorrect(
      tree,
      'apps/my-app/vite.config.ts',
      buildOption,
      buildOptionObject,
      imports,
      plugins,
      testOption,
      testOptionObject,
      '',
      { build: false, test: false, serve: false }
    );
    const appFileContent = tree.read('apps/my-app/vite.config.ts', 'utf-8');
    expect(appFileContent).toMatchSnapshot();
  });

  it('should not do anything if project has everything setup already', () => {
    tree.write('apps/my-app/vite.config.ts', hasEverything);
    ensureViteConfigIsCorrect(
      tree,
      'apps/my-app/vite.config.ts',
      buildOption,
      buildOptionObject,
      imports,
      plugins,
      testOption,
      testOptionObject,
      '',
      { build: true, test: true, serve: true }
    );
    const appFileContent = tree.read('apps/my-app/vite.config.ts', 'utf-8');
    expect(appFileContent).toMatchInlineSnapshot(`
      "
      import { defineConfig } from 'vite';
          import react from '@vitejs/plugin-react';
          import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';
          import dts from 'vite-plugin-dts';
          import { joinPathFragments } from '@nx/devkit';

          export default defineConfig({
            cacheDir: '../../node_modules/.vitest',
            plugins: [dts({ entryRoot: 'src', tsConfigFilePath: joinPathFragments(__dirname, 'tsconfig.lib.json'), skipDiagnostics: true }), react(), nxViteTsPaths(), ],
          
            // Configuration for building your library.
            // See: https://vitejs.dev/guide/build.html#library-mode
            build: {
              lib: {
                // Could also be a dictionary or array of multiple entry points.
                entry: 'src/index.ts',
                name: 'pure-libs-react-vite',
                fileName: 'index',
                // Change this to the formats you want to support.
                // Don't forget to update your package.json as well.
                formats: ['es'],
              },
              rollupOptions: {
                // External packages that should not be bundled into your library.
                external: ['react', 'react-dom', 'react/jsx-runtime'],
              },
            },
          
            test: {
              globals: true,
              environment: 'jsdom',
              include: ['src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],
            },
          });
          "
    `);
  });

  it('should add build option but not update test option if test already setup', () => {
    tree.write('apps/my-app/vite.config.ts', noBuildOptionsHasTestOption);
    ensureViteConfigIsCorrect(
      tree,
      'apps/my-app/vite.config.ts',
      buildOption,
      buildOptionObject,
      imports,
      plugins,
      testOption,
      testOptionObject,
      '',
      { build: false, test: true, serve: true }
    );
    const appFileContent = tree.read('apps/my-app/vite.config.ts', 'utf-8');
    expect(appFileContent).toMatchInlineSnapshot(`
      "import dts from 'vite-plugin-dts';
      import { joinPathFragments } from '@nx/devkit'
      import { defineConfig } from 'vite';
          import react from '@vitejs/plugin-react';
          import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';

          export default defineConfig({
          
            
          // Configuration for building your library.
          // See: https://vitejs.dev/guide/build.html#library-mode
          build: {
            lib: {
              // Could also be a dictionary or array of multiple entry points.
              entry: 'src/index.ts',
              name: 'my-app',
              fileName: 'index',
              // Change this to the formats you want to support.
              // Don't forget to update your package.json as well.
              formats: ['es']
            },
            rollupOptions: {
              // External packages that should not be bundled into your library.
              external: ['react', 'react-dom', 'react/jsx-runtime']
            }
          },cacheDir: '../../node_modules/.vitest',
            plugins: [react(), nxViteTsPaths(), ],

            test: {
              globals: true,
              environment: 'jsdom',
              include: ['src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],
            },

          });
          "
    `);
  });

  it('should update both test and build options - keep existing settings', () => {
    tree.write('apps/my-app/vite.config.ts', someBuildOptionsSomeTestOption);
    ensureViteConfigIsCorrect(
      tree,
      'apps/my-app/vite.config.ts',
      buildOption,
      buildOptionObject,
      imports,
      plugins,
      testOption,
      testOptionObject,
      '',
      { build: false, test: false, serve: true }
    );
    const appFileContent = tree.read('apps/my-app/vite.config.ts', 'utf-8');
    expect(appFileContent).toMatchInlineSnapshot(`
      "import dts from 'vite-plugin-dts';
      import { joinPathFragments } from '@nx/devkit'
      import { defineConfig } from 'vite';
          import react from '@vitejs/plugin-react';
          import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';

          export default defineConfig({
            plugins: [react(), nxViteTsPaths(), ],

            test: {
          'my': 'option',
          'globals': true,
          'environment': "jsdom",
          'include': ["src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}"],
        },

            build: {
          'my': 'option',
          'lib': {"entry":"src/index.ts","name":"my-app","fileName":"index","formats":["es"]},
          'rollupOptions': {"external":["react","react-dom","react/jsx-runtime"]},
        }

          });
          "
    `);
  });
});
