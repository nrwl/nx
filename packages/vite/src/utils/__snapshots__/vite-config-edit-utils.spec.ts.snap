// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ensureViteConfigIsCorrect should add build options if it is using conditional config - do nothing for test 1`] = `
"
    /// <reference types="vitest" />
    import { defineConfig } from 'vite';
    export default defineConfig(({ command, mode, ssrBuild }) => {
      if (command === 'serve') {
        return {
          port: 4200,
          host: 'localhost',
        }
      } else {
        // command === 'build'
        return {
        ...{
          my: 'option',
        },
        ..."\\n    // Configuration for building your library.\\n    // See: https://vitejs.dev/guide/build.html#library-mode\\n    build: {\\n      lib: {\\n        // Could also be a dictionary or array of multiple entry points.\\n        entry: 'src/index.ts',\\n        name: 'my-app',\\n        fileName: 'index',\\n        // Change this to the formats you want to support.\\n        // Don't forget to update your package.json as well.\\n        formats: ['es']\\n      },\\n      rollupOptions: {\\n        // External packages that should not be bundled into your library.\\n        external: ['react', 'react-dom', 'react/jsx-runtime']\\n      }\\n    },"
      }
      }
    })
    "
`;

exports[`ensureViteConfigIsCorrect should not do anything if cannot understand syntax of vite config 1`] = `"console.log('Unknown syntax')"`;
